ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -672,34 +672,52 @@ func (r *RayClusterReconciler) updateStatus(instance *rayiov1alpha1.RayCluster)
 		}
 	}
 
+	if err := r.updateEndpoints(instance); err != nil {
+		return err
+	}
+
+	timeNow := metav1.Now()
+	instance.Status.LastUpdateTime = &timeNow
+	if err := r.Status().Update(context.Background(), instance); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *RayClusterReconciler) updateEndpoints(instance *rayiov1alpha1.RayCluster) error {
+	// TODO: (@scarlet25151) For now there are be several K8s Services for
+	// one RayCluster. We filter for head Services with label selectors and pick the first one.
+	// We may need use the Get method to select by name in the future.
 	rayHeadSvc := corev1.ServiceList{}
-	filterLabels = client.MatchingLabels{
+	filterLabels := client.MatchingLabels{
 		common.RayClusterLabelKey:  instance.Name,
 		common.RayNodeTypeLabelKey: "head",
 	}
-	// TODO: (@scarlet25151) for now there would be several kubernetes serivces related to
-	// one raycluster, we have used the label to select the headservice and pick the first one.
-	// we may need use Get method to select by name.
 	if err := r.List(context.TODO(), &rayHeadSvc, client.InNamespace(instance.Namespace), filterLabels); err != nil {
 		return err
 	}
+
 	if len(rayHeadSvc.Items) != 0 {
 		svc := rayHeadSvc.Items[0]
 		if instance.Status.Endpoints == nil {
 			instance.Status.Endpoints = map[string]string{}
 		}
 		for _, port := range svc.Spec.Ports {
 			if len(port.Name) == 0 {
-				r.Log.Info("updateStatus", "service port name is empty", port)
+				r.Log.Info("updateStatus", "service port's name is empty. Not adding it to RayCluster status.endpoints", port)
 				continue
 			}
-			instance.Status.Endpoints[port.Name] = fmt.Sprintf("%d", port.NodePort)
+			if port.TargetPort.IntVal != 0 {
+				instance.Status.Endpoints[port.Name] = fmt.Sprintf("%d", port.TargetPort.IntVal)
+			} else if port.TargetPort.StrVal != "" {
+				instance.Status.Endpoints[port.Name] = port.TargetPort.StrVal
```

Comment:
Is it OK to leave as port name here or do all endpoint values have to be actual port numbers?

Reply:
/cc @scarlet25151 Please help review this change

Reply:
@davidxia I think it would be good to put all basic head service endpoint here because at this moment all the service would be ready and if there is no port related to it we may logs some error.

Reply:
@scarlet25151 does that mean use actual port numbers? If so, that would require more K8s API calls to get the port number from the Pod spec. Let me know if that's an OK approach.

Reply:
@davidxia I see your point, sorry for the misleading in previous discussion. The endpoint for this place would be some port used to connect to the headpods' k8s service endpoint, for now if the service is using nodePort, we need to use the nodeport so it can be connected correctly. In other case, I think leave as port name here is enough.

Reply:
I see. I'll check if nodePort is "0". If it is, I'll check targetPort and skip adding any endpoint whose value is falsy.

Reply:
@scarlet25151 updated, PTAL 🙏